70_FLIPFluids/gpuRadixSort.h

```diff
@@ -0,0 +1,470 @@
+				};
+				m_device->updateDescriptorSets(std::span(writes, 5), {});
+			}
+			{	// sort: iteration 1 and 3
+				IGPUDescriptorSet::SDescriptorInfo infos[5];
+				infos[0].desc = smart_refctd_ptr(paramsBuffer);
+				infos[0].info.buffer = {.offset = 0, .size = paramsBuffer->getSize()};
+				infos[1].desc = smart_refctd_ptr(tempDataBuffer);
+				infos[1].info.buffer = {.offset = 0, .size = tempDataBuffer->getSize()};
+				infos[2].desc = smart_refctd_ptr(dataBuffer);
+				infos[2].info.buffer = {.offset = 0, .size = dataBuffer->getSize()};			
+				infos[3].desc = smart_refctd_ptr(globalHistogramsBuffer);
+				infos[3].info.buffer = {.offset = 0, .size = globalHistogramsBuffer->getSize()};
+				infos[4].desc = smart_refctd_ptr(partitionHistogramBuffer);
+				infos[4].info.buffer = {.offset = 0, .size = partitionHistogramBuffer->getSize()};
+				IGPUDescriptorSet::SWriteDescriptorSet writes[5] = {
+					{.dstSet = m_radixSortDs[1].get(), .binding = 0, .arrayElement = 0, .count = 1, .info = &infos[0]},
+					{.dstSet = m_radixSortDs[1].get(), .binding = 1, .arrayElement = 0, .count = 1, .info = &infos[1]},
+					{.dstSet = m_radixSortDs[1].get(), .binding = 2, .arrayElement = 0, .count = 1, .info = &infos[2]},
+					{.dstSet = m_radixSortDs[1].get(), .binding = 3, .arrayElement = 0, .count = 1, .info = &infos[3]},
+					{.dstSet = m_radixSortDs[1].get(), .binding = 4, .arrayElement = 0, .count = 1, .info = &infos[4]}
+				};
+				m_device->updateDescriptorSets(std::span(writes, 5), {});
+			}
+		}
+
+		SBufferRange<IGPUBuffer> range;
+		range.buffer = globalHistogramsBuffer;
+		range.size = globalHistogramsBuffer->getSize();
+		cmdbuf->fillBuffer(range, 0ull);
+
+		SMemoryBarrier memBarrier;
+		memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS;
+		memBarrier.srcAccessMask = ACCESS_FLAGS::TRANSFER_WRITE_BIT;
+		memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+		memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+		cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+
+        SSortParams params;
+        params.numElements = numElements;
+        params.numWorkgroups = numWorkgroups;
+        params.numThreadsPerGroup = numThreadsPerGroup;
+
+        for (uint32_t i = 0; i < numIterations; i++)
+        {
+            params.bitShift = 8 * i;
+
+			SBufferRange<IGPUBuffer> range;
+            range.buffer = paramsBuffer;
+            range.size = paramsBuffer->getSize();
+            cmdbuf->updateBuffer(range, &params);
+
+            {
+				SMemoryBarrier memBarrier;
+				memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS;
+				memBarrier.srcAccessMask = ACCESS_FLAGS::TRANSFER_WRITE_BIT;
+				memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+				memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+				cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+            }
```

Comment:
too many barriers, you've just had a barrier from Transfer to Compute because of the `fillBuffer`. They have identical stages, there's nothing in between them that cannot run simultaenously.


This latter barrier just before the dispatch is fine by itself, it would also protect your `fillBuffer`

This would be true even if both varriers were not global and were per-buffer (you'd just do 1 barrier with 2 buffer entries inside).